### PR TITLE
feat: allow partial relPath matches

### PR DESCRIPTION
**Pull request by mkenigs**
The current match-name and match search strategies will find exact matches for relPath (e.g. python310Packages.requests), but they will not return partial matches (e.g. Packages.requests).

Instead:
- Change the match-name and match strategies to no longer return results when an exact relPath is provided. Only match against pname, attrName, and in the case of match, description.
- Add a match-name-or-rel-path search strategy that returns partial matches against relPath. Make this the default search strategy.

### CODE REVIEW

File: cli/flox-rust-sdk/src/models/search.rs [hunk 1]
```diff
@@ -171,10 +182,12 @@ impl Query {
...
                 };
-                if prefer_match_name {
-                    q.match_name = Some(package.to_string());
-                } else {
-                    q.r#match = Some(package.to_string());
+                match search_strategy {
+                    SearchStrategy::Match => q.r#match = Some(package.to_string()),
+                    SearchStrategy::MatchName => q.match_name = Some(package.to_string()),
+                    SearchStrategy::MatchNameOrRelPath => {
+                        q.match_name_or_rel_path = Some(package.to_string())
```
Comment (mkenigs) [inline cli/flox-rust-sdk/src/models/search.rs:189]: I think we should probably use match name whenever there's an `@`, but we can do that in a followup

File: cli/flox-rust-sdk/src/models/search.rs [hunk 2]
```diff
@@ -115,6 +115,15 @@ impl std::fmt::Display for PathOrJson {
     }
 }
 
+#[derive(Clone, Debug, Deserialize, Serialize, Default, PartialEq)]
+#[serde(rename_all = "kebab-case")]
+pub enum SearchStrategy {
+    Match,
+    MatchName,
```
Comment (mkenigs) [inline cli/flox-rust-sdk/src/models/search.rs:122]: Not sure if we should just drop `Match` or `MatchName`, but adding `MatchNameOrRelPath` was pretty boilerplatey so I just went ahead and did it

File: pkgdb/src/pkgdb/pkg-query.cc
Comment (zmitchell) [inline pkgdb/src/pkgdb/pkg-query.cc:1]: @billlevine this file is where you'll be looking when you start working on escaping queries

File: pkgdb/src/search/params.cc [hunk 3]
```diff
@@ -57,7 +57,24 @@ SearchQuery::check() const
...
-        "`partialmatch' and `partialNameMatch' filters "
+        "`partialMatch' and `partialNameMatch' filters "
+        "may not be used together." );
+    }
+  /* `partialMatch' and `partialNameOrRelPathMatch' cannot be used together. */
+  if ( this->partialMatch.has_value()
+       && this->partialNameOrRelPathMatch.has_value() )
+    {
+      throw ParseSearchQueryException(
+        "`partialMatch' and `partialNameOrRelPathMatch' filters "
```
Comment (zmitchell) [inline pkgdb/src/search/params.cc:68]: Blocker: we decided on single quotes or all user-facing text.

**Review by zmitchell — CHANGES_REQUESTED** [review]
LGTM other than the quotes issue

**Review by zmitchell — APPROVED** [review]

File: pkgdb/src/search/params.cc (hunk 3 above)
Comment (mkenigs) [reply pkgdb/src/search/params.cc]: Oops I may have replaced every backtick in the file 😉

**Pull request merged**